File: src/runtime/client/protocol.ts
// The wire contract shared by SocketRuntimeServer and SocketRuntimeClient: the
// JSON-RPC method/event names and error (de)serialization. Keeping the names in
// one place stops the two sides from drifting. The Go client mirrors these.
// See docs/go-tui-migration-plan.md §4.

// Client → server requests (and one fire-and-forget notification, recordInput).
export const RpcMethod = {
  GetConfig: "getConfig",
  Reload: "reload",
  Execute: "execute",
  RecordInput: "recordInput",
  ListAiSessions: "cache/listAiSessions",
  ListApiEndpoints: "cache/listApiEndpoints",
  ListTraceCalls: "cache/listTraceCalls",
  ClearCache: "cache/clear",
  AiStart: "ai/start",
  AiPrompt: "ai/prompt",
  AiRespondPermission: "ai/respondPermission",
  AiStop: "ai/stop",
} as const

// Server → client notifications.
export const RpcEvent = {
  SessionUpdate: "event/sessionUpdate",
  ConversationUpdate: "event/conversationUpdate",
  PermissionRequest: "event/permissionRequest",
  SessionStarted: "event/sessionStarted",
  SessionStopped: "event/sessionStopped",
  SessionError: "event/sessionError",
  ExternalEvent: "event/externalEvent",
  Error: "event/error",
} as const

// Errors can't cross JSON as Error instances; carry the message (and name) and
// rebuild an Error on the far side.
export type SerializedError = { message: string; name?: string }

export const serializeError = (error: unknown): SerializedError => {
  if (error instanceof Error) {
    return { message: error.message, name: error.name }
  }
  return { message: String(error) }
}

export const deserializeError = (value: SerializedError): Error => {
  const error = new Error(value.message)
  if (value.name) {
    error.name = value.name
  }
  return error
}

File: tui/internal/runtime/protocol.go
// Package runtime is the Go client for the ntee-r1quest runtime server: it
// speaks the same JSON-RPC protocol as the TS SocketRuntimeClient
// (src/runtime/client) over a Unix-domain socket. Method/event names and DTOs
// mirror src/runtime/client/protocol.ts and types.ts exactly.
// See docs/go-tui-migration-plan.md §4.
package runtime

// Client → server requests (RecordInput / AiStop are sent as notifications).
const (
	MethodGetConfig           = "getConfig"
	MethodReload              = "reload"
	MethodExecute             = "execute"
	MethodRecordInput         = "recordInput"
	MethodSuggestInputs       = "cache/suggestInputs"
	MethodListAiSessions      = "cache/listAiSessions"
	MethodListApiEndpoints    = "cache/listApiEndpoints"
	MethodListTraceCalls      = "cache/listTraceCalls"
	MethodClearCache          = "cache/clear"
	MethodSnapshotPut         = "snapshot/put"
	MethodSnapshotGet         = "snapshot/get"
	MethodSnapshotList        = "snapshot/list"
	MethodSnapshotDelete      = "snapshot/delete"
	MethodAiStart             = "ai/start"
	MethodAiPrompt            = "ai/prompt"
	MethodAiRespondPermission = "ai/respondPermission"
	MethodAiStop              = "ai/stop"
)

// Server → client notifications.
const (
	EventSessionUpdate      = "event/sessionUpdate"
	EventConversationUpdate = "event/conversationUpdate"
	EventPermissionRequest  = "event/permissionRequest"
	EventSessionStarted     = "event/sessionStarted"
	EventSessionStopped     = "event/sessionStopped"
	EventSessionError       = "event/sessionError"
	EventExternalEvent      = "event/externalEvent"
	EventError              = "event/error"
)
